### PR TITLE
Piped communication on Windows

### PR DESCRIPTION
**Pull request by TGWDB**
Piped inter-process communication for Windows.

**Old text below.**

This is a PR to test CI and see how the piped communication progresses on Windows. This started based on #6151 and so will need to either wait for that to be merged and rebased, or supersede that PR.

Note that this is part of the larger SMT plan available at #6134 .

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/piped_process.cpp
Comment (kroening) [inline src/util/piped_process.cpp:135]: What access rights does this security descriptor grant? The default one for named pipes gives read access to the EVERYONE group.

File: src/util/piped_process.cpp
Comment (kroening) [inline src/util/piped_process.cpp]: What happens when there are multiple instances of CBMC running simultaneously? Do they all try to talk over the same named pipe?

File: src/util/piped_process.cpp
Comment (kroening) [inline src/util/piped_process.cpp]: This will require quoting. Use `quote_windows_arg` from util/run.cpp

File: src/util/piped_process.cpp
Comment (kroening) [inline src/util/piped_process.cpp:432]: There is no better option than polling? This will heat a lot of CPUs!

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp:135]: If I read the documentation [here](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/aa379560(v=vs.85)) correctly then by defining the security attributes for `null` for `lpSecurityDescriptor` (`sec_attr` approx line 30 of `src/util/piped_process.cpp`) this sets the security to be the default for the current session access token. In particular this is NOT available to all users.

That said, I'll add comment about this to the code.

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Fixed with use of random number.

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp:135]: Added comment to code as well.

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: I'd prefer to leave this to the calling code to properly ensure the handling of quotations (also this breaks unit tests and makes for messy dependencies for now).

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp:432]: I haven't found evidence of one that works nicely with our asynchronous reading from named pipes. The options available for waiting appear to work with blocking pipes, but this does not work with our interactive approach (at least at the moment).

If someone who knows Windows better can suggest an alternative I'll look into it.

File: src/util/piped_process.cpp
Comment (kroening) [reply src/util/piped_process.cpp]: `rand` isn't random, but produces a deterministic sequence unless you change the seed.
Consider using something you know to be unique, say a process ID.

File: src/util/piped_process.cpp
Comment (kroening) [reply src/util/piped_process.cpp]: This is a no-go, since it has security implications.

File: src/util/piped_process.cpp
Comment (kroening) [reply src/util/piped_process.cpp:432]: I'd say this is a blocker for using this.

File: src/util/piped_process.cpp
Comment (kroening) [inline src/util/piped_process.cpp]: This should use `CreateProcessW`, and `widen(cmdline)`

File: src/util/piped_process.cpp
Comment (kroening) [inline src/util/piped_process.cpp]: Does this clean up the named pipe?

Comment (TGWDB) [reply src/util/piped_process.cpp]: According to the docs when all handles are closed/released then Windows cleans up the pipe. I've pushed some more aggressive killing and disconnecting of the child process, but it probably makes no difference in practice.

File: src/util/piped_process.cpp
Comment (kroening) [reply src/util/piped_process.cpp:432]: `MsgWaitForMultipleObjectsEx` might do the right thing.

Comment (TGWDB) [reply src/util/piped_process.cpp:432]: `WaitForMultipleObjects` (and similar ones) rely upon the events they wait on being synchronous. This (so far in testing and from what I've read) means we have to wait for the child process to finish before we can read from the pipe. I saw some example code that used this approach with asynchronous pipes, but then used a busy loop with `peek` in the same way I've implemented currently.

I'll keep looking, and play around more now I've sorted out most of the other quirks.

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Added quoting here.

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: It was seeded, but process ID is better - done.

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Done, along with quoting.

File: src/util/piped_process.cpp
Comment (kroening) [inline src/util/piped_process.cpp]: You need to add one `WCHAR` to the size of the allocated buffer for the terminating zero.
To avoid this problem altogether, use `_wcsdup`.

File: src/util/piped_process.cpp
Comment (kroening) [inline src/util/piped_process.cpp]: No need for the `strdup`, and the subsequent `free`.

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Done

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: The child process wants to be able to modify the arguments (i.e. non `const`) while the `commandvec`  with `c_str()`can only provide `const` instances.

Comment (kroening) [reply src/util/piped_process.cpp]: This memory is not shared with the child; the child gets a copy. `const` is fine.

Comment (TGWDB) [reply src/util/piped_process.cpp]: Fixed. :)

Comment (kroening) [reply src/util/piped_process.cpp]: Please add the `const` to the type of `args`. There is then no need for the `const_cast`.

**Comment by TGWDB** [comment]
Now rebased on develop after merging of #6151 .

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp:432]: See comment/discussion at the top of the file for more on why this is done this way.

File: src/util/piped_process.cpp
Comment (thomasspriggs) [inline src/util/piped_process.cpp]: :pick: I suggest declaring success where it is used - `const BOOL success = CreateProcessW(`

File: src/util/piped_process.cpp
Comment (thomasspriggs) [inline src/util/piped_process.cpp]: :pick: I suggest splitting the lines of code which build `cmdline` out into a separate function and then declaring the variable holding the string returned as `const`.

File: src/util/piped_process.cpp
Comment (thomasspriggs) [inline src/util/piped_process.cpp:342]: :pick: I would have kept this rename in a separate commit, in order to keep the commits smaller and restricted to a single change. I consider this something to consider for next time, rather than something worth editing the commit history to fix.

File: src/util/piped_process.cpp
Comment (thomasspriggs) [inline src/util/piped_process.cpp]: :no_entry_sign: :question: Are we violating the C++ rule of three which says that if we define a custom destructor then we also need to define the copy constructor and copy assignment operator? I wasn't sure if these other operations were implicitly deleted. If we violate this rule then we are at risk of copying this class and then freeing the resources it handles multiple times. The simplest fix for this issue could be to explicitly delete the other 2, to prevent copying. Alternatively the code in the destructor could be moved into custom deleters for the process info and handles using `std::unique_ptr` / `std::shared_ptr` pointers pointing to the managed resources.

File: src/util/piped_process.cpp
Comment (thomasspriggs) [inline src/util/piped_process.cpp]: :bulb: I suggest using `const std::string in_name = base_name + "\\IN";` rather than using `.append`. This allows us to both make the variable const and use a slightly more succinct syntax.

File: src/util/piped_process.cpp
Comment (thomasspriggs) [inline src/util/piped_process.cpp:180]: :no_entry_sign: :question: Why is the usage of the STOPPED state in the non-Windows version being removed as part of the implementation of the Windows version?

File: src/util/piped_process.cpp
Comment (thomasspriggs) [inline src/util/piped_process.cpp:423]: :pick: The magic value of `-1` for `timeout` is set due to the unix API being used. I suggest the "unwrap" is moved into the non-Windows code and this line be written as `while(!wait_time || waited_time >= *wait_time)`

File: src/util/piped_process.h
Comment (thomasspriggs) [inline src/util/piped_process.h]: :no_entry_sign: Including `windows.h` in this header file exposes the entire windows API to any other file which includes this one when building on Windows. This has the potential to make it more difficult for contributors working on Windows to write platform independent code and could also increase build times. This issue could be worked around by forward declaring `PROCESS_INFORMATION` and `HANDLE` and making the member fields pointers to these types, or by using the "Pointer to implementation" / "pImpl" technique to only declare fields of these types in the `.cpp` file.

File: unit/util/piped_process.cpp
Comment (thomasspriggs) [inline unit/util/piped_process.cpp]: :pick: This should probably be done using C++11's `std::chrono::steady_clock` from the `<chrono>` header, so as to avoid unnecessarily making the test operating system specific.

**Review by thomasspriggs — CHANGES_REQUESTED** [review]
Great to see some more progress on this, but there are some :no_entry_sign: issues which we should address before merging this.

File: src/util/piped_process.cpp
Comment (peterschrammel) [inline src/util/piped_process.cpp]: pipe

File: src/util/piped_process.cpp
Comment (peterschrammel) [inline src/util/piped_process.cpp]: behaviour

File: src/util/piped_process.cpp
Comment (peterschrammel) [inline src/util/piped_process.cpp]: much?

File: src/util/piped_process.cpp
Comment (peterschrammel) [inline src/util/piped_process.cpp:68]: Would something like this work instead of sleep or is this no better but just with more overhead?
```
do {
  // check whether there's something new
} while (WaitForSingleObject(hEvent, 100) == WAIT_TIMEOUT)
```

File: src/util/piped_process.cpp
Comment (peterschrammel) [inline src/util/piped_process.cpp]: comment should be one line further down

File: src/util/piped_process.cpp
Comment (peterschrammel) [inline src/util/piped_process.cpp]: This is the timeout to detect that the other side hasn't connected, I guess?

File: src/util/piped_process.cpp
Comment (peterschrammel) [inline src/util/piped_process.cpp]: I think we have exception types in `util/exception_util.h` that are more suitable.

File: src/util/piped_process.cpp
Comment (peterschrammel) [inline src/util/piped_process.cpp]: Would it be possible to split up this mega-function a bit?

File: src/util/piped_process.cpp
Comment (tautschnig) [reply src/util/piped_process.cpp]: Also: MacOS

File: src/util/piped_process.cpp
Comment (tautschnig) [inline src/util/piped_process.cpp]: I recognise this was here before, but any insight as to why this isn't being passed as a reference? The `const` doesn't seem terribly useful at present.

File: src/util/piped_process.cpp
Comment (tautschnig) [inline src/util/piped_process.cpp]: This is generic code in `util`, I don't think it should hard-code `SMT2`.

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp:68]: Those don't work on non-blocking pipes, they immediately return and so would not actually sleep.

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Fixed

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Fixed

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Fixed

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp:180]: Old behaviour preserved, added comment about to check if we really need this state (TODO)

File: src/util/piped_process.h
Comment (TGWDB) [reply src/util/piped_process.h]: Fixed.

File: unit/util/piped_process.cpp
Comment (TGWDB) [reply unit/util/piped_process.cpp]: Fixed

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Fixed

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Both fixed

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Fixed

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Fixed

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Fixed

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Fixed

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: No insight, but fixed now.

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Split out a small helper function, but the function is pretty straightforward, just many lines due to many arguments to function calls to create Windows processes/pipes (also lots of commends/explanation).

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp:180]: Removed excess states (in recent separate commit).

File: src/util/piped_process.cpp
Comment (TGWDB) [reply src/util/piped_process.cpp]: Fixed

File: src/util/piped_process.cpp
Comment (thomasspriggs) [inline src/util/piped_process.cpp]: :bulb: I would declare this as `static std::wstring process_windows_args(const std::vector<std::string> commandvec)`. Adding the `static` tells the compiler that this function has internal linkage only. Meaning that it can only be called from within this `.cpp` file. This informs the reader that this function is an internal implementation detail, not part of the external interface.

File: src/util/piped_process.cpp
Comment (thomasspriggs) [inline src/util/piped_process.cpp]: :pick: I suggest reusing this comment as doxygen just above the definition of the new function.

File: src/util/piped_process.cpp
Comment (thomasspriggs) [inline src/util/piped_process.cpp]: :pick: `process` is a little non-descriptive. Maybe `prepare_windows_command_line` would be better.

File: src/util/piped_process.cpp
Comment (thomasspriggs) [inline src/util/piped_process.cpp]: :no_entry_sign: Adding that `*` passes by pointer, not by reference. To pass by reference you should use `&`. Passing by pointer is an inferior solution as it allows `nullptr` to be passed as an argument, whereas a reference is non-nullable (as is passing by value). This issue should be fixed before this PR is merged, because it affects the way the constructor is used. So if we leave resolving it later then we may need to fix additional usages of the constructor built on top of this PR.

Comment (thomasspriggs) [reply src/util/piped_process.cpp]: Looks to be fixed.

**Review by thomasspriggs — APPROVED** [review]
:+1:

File: src/util/piped_process.cpp
Comment (peterschrammel) [reply src/util/piped_process.cpp]: Any insights on this? Would be good to explain to save follow-ups the time for investigating how the API works.

File: unit/util/piped_process.cpp
Comment (peterschrammel) [inline unit/util/piped_process.cpp]:
```suggestion
  const std::string expected_error("'abcde' is not recognized");
```

File: unit/util/piped_process.cpp
Comment (peterschrammel) [inline unit/util/piped_process.cpp]:
```suggestion
  // Currently not working under Linux/MacOS?!
```

File: unit/util/piped_process.cpp
Comment (peterschrammel) [inline unit/util/piped_process.cpp]: Put in `#if 0` rather than commenting.

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**
